File: SourceAFIS/General/DetailLogger.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Reflection;
using SourceAFIS.Meta;
using SourceAFIS.Dummy;

namespace SourceAFIS.General
{
    public sealed class DetailLogger
    {
        public abstract class Hook
        {
            public abstract bool IsActive { get; }
            public abstract void Log(object data);
            public abstract void Log(string part, object data);
        }

        class NullHook : Hook
        {
            public override bool IsActive { get { return false; } }
            public override void Log(object data) { }
            public override void Log(string part, object data) { }
        }

        class ActiveHook : Hook, ICloneable
        {
            public DetailLogger Logger;
            public string Path;

            public ActiveHook(DetailLogger logger, string path)
            {
                Logger = logger;
                Path = path;
            }

            public override bool IsActive { get { return true; } }

            public override void Log(object data)
            {
                Logger.Log(Path, data);
            }

            public override void Log(string part, object data)
            {
                Logger.Log(Path + "." + part, data);
            }

            public ActiveHook Clone() { return new ActiveHook(Logger, Path); }
            object ICloneable.Clone() { return Clone(); }
        }

        public class LogData
        {
            Dictionary<string, List<object>> History = new Dictionary<string, List<object>>();

            public void Append(string path, object data)
            {
                if (!History.ContainsKey(path))
                    History[path] = new List<object>();
                History[path].Add(data);
            }

            public object Retrieve(string path)
            {
                lock (this)
                    return Retrieve(path, 0);
            }

            public object Retrieve(string path, int index)
            {
                lock (this)
                {
                    if (History.ContainsKey(path) && index < History[path].Count)
                        return History[path][index];
                    else
                        return null;
                }
            }
        }

        public static readonly Hook Null = new NullHook();

        [ThreadStatic]
        static string ThreadName;

        LogData CurrentLog = new LogData();

        public LogData PopLog()
        {
            LogData result = CurrentLog;
            CurrentLog = new LogData();
            return result;
        }

        public void Attach(ObjectTree tree)
        {
            foreach (object reference in tree.GetAllObjects())
            {
                FieldInfo field = reference.GetType().GetField("Logger");
                if (field != null)
                    field.SetValue(reference, new ActiveHook(this, tree.GetPath(reference)));
            }
        }

        public void Log(string path, object data)
        {
            lock (this)
            {
                object logged;
                if (data is ICloneable)
                    logged = (data as ICloneable).Clone();
                else
                    logged = data;
                string qualifiedPath = path + GetThreadName();
                CurrentLog.Append(qualifiedPath, logged);
            }
        }

        static string GetThreadName()
        {
            if (ThreadName != null)
                return "[" + ThreadName + "]";
            else
                return "";
        }

        public static void RunInContext(string name, Action task)
        {
            ThreadName = name;
            try
            {
                task();
            }
            finally
            {
                ThreadName = null;
            }
        }

        public static void CopyHooks(object original, object copy)
        {
            ObjectTree originalTree = new ObjectTree(original);
            ObjectTree copyTree = new ObjectTree(copy);
            foreach (object originalReference in originalTree.GetAllObjects())
            {
                FieldInfo field = originalReference.GetType().GetField("Logger");
                if (field != null)
                {
                    object hook = field.GetValue(originalReference);
                    if (hook is ActiveHook)
                    {
                        string path = originalTree.GetPath(originalReference);
                        if (copyTree.Contains(path))
                        {
                            object copyReference = copyTree.GetObject(path);
                            field.SetValue(copyReference, (hook as ActiveHook).Clone());
                        }
                    }
                }
            }
        }
    }
}

File: java/sourceafis/src/sourceafis/general/DetailLogger.java
package sourceafis.general;

import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.ArrayList;
import java.util.HashMap;

import sourceafis.meta.ObjectTree;

public class DetailLogger {
	public static final Hook off = new NullHook();
	private LogData currentLog = new LogData();
	
	public void attach(ObjectTree tree) {
		try {
			for (Object reference : tree.getAllObjects())
				for (Field field : reference.getClass().getFields())
					if (field.getName().equals("logger"))
						field.set(reference, new ActiveHook(tree.getPath(reference)));
		} catch (IllegalAccessException e) {
			throw new RuntimeException(e);
		}
	}
	
	public synchronized void log(String path, Object data) {
		Object logged = data;
		try {
			for (Method clone : data.getClass().getMethods())
				if (clone.getName().equals("clone") && Modifier.isPublic(clone.getModifiers()))
					logged = clone.invoke(data);
		} catch (InvocationTargetException e) {
			throw new RuntimeException(e);
		} catch (IllegalAccessException e) {
			throw new RuntimeException(e);
		}
		currentLog.append(path, logged);
	}
	
	public LogData popLog() {
		LogData result = currentLog;
		currentLog = new LogData();
		return result;
	}
	
	public static void copyHooks(Object original, Object copy) {
		ObjectTree originalTree = new ObjectTree(original);
		ObjectTree copyTree = new ObjectTree(copy);
		for (Object originalRef : originalTree.getAllObjects())
			for (Field field : originalRef.getClass().getFields())
				if (field.getName().equals("logger")) {
					Object copyRef = copyTree.getObject(originalTree.getPath(originalRef));
					try {
						field.set(copyRef, field.get(originalRef));
					} catch (IllegalAccessException e) {
						throw new RuntimeException(e);
					}
				}
	}

	public static abstract class Hook {
        public abstract boolean isActive();
        public abstract void log(Object data);
        public abstract void log(String part, Object data);
	}
	
	private static class NullHook extends Hook {
        public boolean isActive() { return false; }
        public void log(Object data) {}
        public void log(String part, Object data) {}
	}
	
	private class ActiveHook extends Hook implements Cloneable {
		private String path;
		
		public ActiveHook(String p) { path = p; }
        public boolean isActive() { return true; }
        public void log(Object data) { DetailLogger.this.log(path, data); }
        public void log(String part, Object data) {
        	DetailLogger.this.log(path + (path != "" ? "." : "") + part, data);
    	}
	}
	
	public class LogData {
		private HashMap<String, ArrayList<Object>> history = new HashMap<String, ArrayList<Object>>();
		
		public void append(String path, Object data) {
			if (!history.containsKey(path))
				history.put(path, new ArrayList<Object>());
			history.get(path).add(data);
		}
		
		public int count(String path) {
			if (history.containsKey(path))
				return history.get(path).size();
			else
				return 0;
		}
		
		public Object retrieve(String path, int index) {
			if (history.containsKey(path) && index < history.get(path).size())
				return history.get(path).get(index);
			else
				return null;
		}
		
		public Object retrieve(String path) { return retrieve(path, 0); }
	}
}
